packages/tpetra/core/src/Tpetra_Experimental_BlockCrsMatrix_def.hpp

```diff
@@ -2929,303 +3020,326 @@ class GetLocalDiagCopy {
+             importsByteHost, 
+             offval, numBytes, numEnt,
+             numBytesPerValue, blockSize);
+
+          if (numBytes != numBytesOut) {
+            errorDuringUnpack() = true; 
+            if (debug) {
+              std::ostream& err = this->markLocalErrorAndGetStream ();
+              err << prefix 
+                  << "At i = " << i << ", numBytes = " << numBytes
+                  << " != numBytesOut = " << numBytesOut << "."
+                  << std::endl;
+            }
+          }
 
-      if (static_cast<LO> (numEnt) != numCombd) {
-        errorDuringUnpack = true;
-#ifdef HAVE_TPETRA_DEBUG
-        std::ostream& err = this->markLocalErrorAndGetStream ();
-        err << prefix << "At i = " << i << ", numEnt = " << numEnt
-            << " != numCombd = " << numCombd << ".";
-#endif // HAVE_TPETRA_DEBUG
-        continue;
-      }
+          // Convert incoming global indices to local indices.
+          for (size_t k = 0; k < numEnt; ++k) {
+            lidsOut(k) = tgtColMap.getLocalElement (gidsOut(k));
+            if (lidsOut(k) == Teuchos::OrdinalTraits<LO>::invalid ()) {
+              errorDuringUnpack() = true; 
+              if (debug) {
+                std::ostream& err = this->markLocalErrorAndGetStream ();
+                err << prefix 
+                    << "At i = " << i << ", GID " << gidsOut(k)
+                    << " is not owned by the calling process." 
+                    << std::endl;
+              }
+            }
+          }
+          
+          // Combine the incoming data with the matrix's current data.
+          LO numCombd = 0;
+          const LO* const lidsRaw = const_cast<const LO*> (lidsOut.data ());
+          const Scalar* const valsRaw =
+            reinterpret_cast<const Scalar*> (const_cast<const impl_scalar_type*> (valsOut.data ()));
+          if (combineMode == ADD) {
+            numCombd = this->sumIntoLocalValues (lclRow, lidsRaw, valsRaw, numEnt);
+          } else if (combineMode == INSERT || combineMode == REPLACE) {
+            numCombd = this->replaceLocalValues (lclRow, lidsRaw, valsRaw, numEnt);
+          } else if (combineMode == ABSMAX) {
+            numCombd = this->absMaxLocalValues (lclRow, lidsRaw, valsRaw, numEnt);
+          }
 
-      // Don't update offset until current LID has succeeded.
-      offset += numBytes;
-    } // for each import LID i
+          if (static_cast<LO> (numEnt) != numCombd) {
+            errorDuringUnpack() = true; 
+            if (debug) {
+            std::ostream& err = this->markLocalErrorAndGetStream ();
+            err << prefix 
+                << "At i = " << i << ", numEnt = " << numEnt
+                << " != numCombd = " << numCombd << "."
+                << std::endl;
+            }
+          }
+        }); // for each import LID i
+    }
 
-    if (errorDuringUnpack) {
+    if (errorDuringUnpack()) {
       std::ostream& err = this->markLocalErrorAndGetStream ();
```

Comment:
Regarding the above thread safety issue of `markLocalErrorAndGetStream()`: Since you already call the latter here, after the parallel loop(s), all you should need to do is take those calls to `markLocalErrorAndGetStream()` out of the loops.